### PR TITLE
[CINFRA-818] Index creation should go through the replicated log

### PR DESCRIPTION
**Pull request by apetenchea**
### Scope & Purpose

This PR introduces the _CreateIndex_ operation in replication2 databases. It allows for index creation to be controlled by the replicated log leader and passed to the followers subsequently.

**Remarks**
- For unique indexes, the leader first builds the index itself and only then replicates it. Other indexes don't have this limitation.
- The maintenance has been patched such that it only affects replication2 leaders. Followers are ignored, they only become aware of new indexes through the replicated log.

**Todos**
- Creating indexes during snapshot transfer - this is a bigger topic, will be handled in different PR.
- I added basic integration tests. In the near future I hope to add some extra unit tests.

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [x] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.11: *(Please link PR)*
  - [ ] Backport for 3.10: *(Please link PR)*

### CODE REVIEW

File: arangod/Replication2/StateMachines/Document/DocumentStateIndexHandler.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+#include <set>
+#include <shared_mutex>
+
+struct TRI_vocbase_t;
+
+namespace arangodb {
+class MaintenanceFeature;
+
+namespace replication2::replicated_state::document {
+struct IDocumentStateIndexHandler {
```
Comment (mchacki) [inline arangod/Replication2/StateMachines/Document/DocumentStateIndexHandler.h:39]: I do not See this struct in Action anywhere, it is also not implemented (at least not checked in)
A method with same signature is implemented on the DocumentStateShardHandler.
Was this intentional, is there still a last step to go? Was this file temporary and not fully removed?

**Review by mchacki — CHANGES_REQUESTED** [review]
The workflow and logic implemented looks good :+1:

I have only some comments on the API layout.
It also seems there was either one step (move code or cleanup) missing

File: arangod/Replication2/StateMachines/Document/DocumentStateIndexHandler.h (hunk 1 above)
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentStateIndexHandler.h]: At first I created this class to handle indexes, but then I realized that indexes are tightly coupled to shards. I moved the functionality to `DocumentStateShardHandler` and removed the .cpp file, but forgot about the header. Thanks for noticing!
Fixed in https://github.com/arangodb/arangodb/pull/19699/commits/101ef1fc6755e2a3bc69dd4f072b76f6be4c98d8

File: arangod/Replication2/StateMachines/Document/DocumentLeaderState.h [hunk 2]
```diff
@@ -74,6 +79,11 @@ struct DocumentLeaderState
   auto dropShard(ShardID shard, CollectionID collectionId)
       -> futures::Future<Result>;
 
+  auto createIndex(LogicalCollection& col, VPackSlice indexInfo,
+                   std::shared_ptr<VPackBuilder> output,
```
[the remark this answers is not in the document]
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentLeaderState.h]: Better clarity is always a plus. Fixed in https://github.com/arangodb/arangodb/pull/19699/commits/27b6eac5dc679ddc236c6173e2f746dcbbe8674b

File: arangod/Replication2/StateMachines/Document/MaintenanceActionExecutor.h [hunk 3]
```diff
@@ -44,13 +45,19 @@ struct IMaintenanceActionExecutor {
       ShardID shard, CollectionID collection,
       std::shared_ptr<VPackBuilder> properties, std::string followersToDrop)
       -> Result = 0;
+  virtual auto executeCreateIndex(
+      ShardID shard, std::shared_ptr<VPackBuilder> const& properties,
+      std::shared_ptr<VPackBuilder> const& output,
```
[the remark this answers is not in the document]
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/MaintenanceActionExecutor.h]: Addressed in https://github.com/arangodb/arangodb/pull/19699/commits/27b6eac5dc679ddc236c6173e2f746dcbbe8674b

File: tests/js/server/replication2/replication2-replicated-state-document-store.js [hunk 4]
```diff
@@ -956,6 +953,7 @@ const replicatedStateSnapshotTransferSuite = function () {
  * This test suite checks the correctness of a DocumentState shard-related operations.
  */
 const replicatedStateDocumentShardsSuite = function () {
+  var collection = null;
```
[the remark this answers is not in the document]
Comment (apetenchea) [reply tests/js/server/replication2/replication2-replicated-state-document-store.js]: Addressed in https://github.com/arangodb/arangodb/pull/19699/commits/10dcb65ec1d5b970414e9116b601cbdbd152eb13

File: arangod/Cluster/EnsureIndex.h [hunk 5]
```diff
@@ -46,10 +46,11 @@ class EnsureIndex : public ActionBase {
   virtual arangodb::Result setProgress(double d) override final;
   virtual bool first() override final;
 
+  static void indexCreationLogging(VPackBuilder& index);
```
Comment (mchacki) [inline arangod/Cluster/EnsureIndex.h:49]: I think this should either be `VPackBuilder const&` or better `VPackSlice`
This method is not supposed to modify the Builder

**Review by mchacki — APPROVED** [review]
LGTM.

Just some very minor nit picking comment, happy to have that ignored

File: arangod/Cluster/EnsureIndex.h (hunk 5 above)
Comment (apetenchea) [reply arangod/Cluster/EnsureIndex.h]: Using `VPackSlice`: https://github.com/arangodb/arangodb/pull/19699/commits/4e4c2130d309787e698d2d0fc261445f047e27c1

**Pull request merged**